File: src/test/java/no/nav/syfo/rest/ressurser/NaermesteLederRessursTest.java
package no.nav.syfo.rest.ressurser;

import no.nav.security.oidc.context.OIDCRequestContextHolder;
import no.nav.syfo.LocalApplication;
import no.nav.syfo.model.NaermesteLederStatus;
import no.nav.syfo.narmesteleder.Naermesteleder;
import no.nav.syfo.narmesteleder.NarmesteLederConsumer;
import no.nav.syfo.rest.domain.RSNaermesteLeder;
import no.nav.syfo.services.AktoerService;
import no.nav.syfo.services.TilgangskontrollService;
import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.junit4.SpringRunner;

import javax.inject.Inject;
import java.util.Optional;

import static no.nav.syfo.testhelper.OidcTestHelper.loggInnBruker;
import static no.nav.syfo.testhelper.OidcTestHelper.loggUtAlle;
import static no.nav.syfo.testhelper.UserConstants.*;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@RunWith(SpringRunner.class)
@SpringBootTest(classes = LocalApplication.class)
@DirtiesContext
public class NaermesteLederRessursTest {

    @Inject
    private NaermestelederRessurs naermestelederRessurs;

    @Inject
    public OIDCRequestContextHolder oidcRequestContextHolder;
    @Inject
    private AktoerService aktoerService;

    @MockBean
    private TilgangskontrollService tilgangskontrollService;
    @MockBean
    private NarmesteLederConsumer narmesteLederConsumer;

    @After
    public void tearDown() {
        loggUtAlle(oidcRequestContextHolder);
    }

    @Test
    public void hentNaermesteLederSuccess() {
        loggInnBruker(oidcRequestContextHolder, LEDER_FNR);

        when(tilgangskontrollService.sporOmNoenAndreEnnSegSelvEllerEgneAnsatte(LEDER_FNR, ARBEIDSTAKER_FNR)).thenReturn(false);

        Naermesteleder leder = new Naermesteleder()
                .naermesteLederAktoerId(LEDER_AKTORID)
                .orgnummer(VIRKSOMHETSNUMMER)
                .naermesteLederStatus(new NaermesteLederStatus()
                        .erAktiv(true));
        when(narmesteLederConsumer.narmesteLeder(anyString(), anyString())).thenReturn(Optional.of(leder));

        RSNaermesteLeder rsNaermesteLeder = naermestelederRessurs.hentNaermesteLeder(ARBEIDSTAKER_FNR, VIRKSOMHETSNUMMER);

        assertEquals(LEDER_FNR, rsNaermesteLeder.fnr);
    }
}

File: src/test/kotlin/no/nav/syfo/narmesteleder/NaermesteLederRessursTest.kt
package no.nav.syfo.narmesteleder

import io.mockk.*
import no.nav.security.token.support.core.context.TokenValidationContextHolder
import no.nav.syfo.consumer.aktorregister.AktorregisterConsumer
import no.nav.syfo.metric.Metric
import no.nav.syfo.narmesteleder.consumer.*
import no.nav.syfo.narmesteleder.controller.NaermestelederController
import no.nav.syfo.testhelper.OidcTestHelper.getValidationContext
import no.nav.syfo.testhelper.UserConstants
import no.nav.syfo.tilgang.TilgangskontrollService
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.time.LocalDate

@ExtendWith(SpringExtension::class)
class NaermesteLederRessursTest {
    private val aktorregisterConsumer: AktorregisterConsumer = mockk()

    private val oidcRequestContextHolder: TokenValidationContextHolder = mockk()

    private val tilgangskontrollService: TilgangskontrollService = mockk()

    private val narmesteLederConsumer: NarmesteLederConsumer = mockk()

    private val metric: Metric = mockk()

    private val naermestelederRessurs = NaermestelederController(
        metric,
        oidcRequestContextHolder,
        tilgangskontrollService,
        aktorregisterConsumer,
        narmesteLederConsumer
    )

    @Test
    fun hentNaermesteLederSuccess() {
        every { oidcRequestContextHolder.tokenValidationContext }.returns(
            getValidationContext(UserConstants.LEDER_FNR)
        )
        every { aktorregisterConsumer.hentFnrForAktor(UserConstants.LEDER_AKTORID) }.returns(UserConstants.LEDER_FNR)
        every { metric.countEndpointRequest(any()) } just Runs
        every { tilgangskontrollService.sporOmNoenAndreEnnSegSelvEllerEgneAnsatte(UserConstants.LEDER_FNR, UserConstants.ARBEIDSTAKER_FNR) }.returns(false)
        val leder = Naermesteleder(
            naermesteLederId = 0L,
            naermesteLederAktoerId = UserConstants.LEDER_AKTORID,
            naermesteLederStatus = NaermesteLederStatus(
                erAktiv = true,
                aktivFom = LocalDate.now().minusDays(1),
                aktivTom = null
            ),
            orgnummer = UserConstants.VIRKSOMHETSNUMMER,
            navn = "",
            epost = null,
            mobil = null
        )
        every { narmesteLederConsumer.narmesteLeder(any(), any()) }.returns(leder)
        val rsNaermesteLeder = naermestelederRessurs.hentNaermesteLeder(UserConstants.ARBEIDSTAKER_FNR, UserConstants.VIRKSOMHETSNUMMER)
        assertEquals(UserConstants.LEDER_FNR, rsNaermesteLeder.fnr)
    }
}
